### PR TITLE
Add meson support

### PR DESCRIPTION
**Pull request by jheaff1**
This PR adds [Meson](https://mesonbuild.com/) support to rules_foreign_cc. 

This PR depends on #987 

This PR is rather large, I tried to make the commits atomic but I can split up the PR into smaller ones  if need be.

### CODE REVIEW

**Comment by jheaff1** [comment]
@jsharpe This PR is now ready for review. Apologies for it being so long, its the culmination of 6 months work!

**Comment by jsharpe** [comment]
I could have sworn I put a reply on this PR but obviously didn't hit comment!
I think a number of these changes are useful in their own rights so I've started creating PRs to merge them as separate PRs; hope you don't mind; once the framework changes are broken out then I think the meson changes can go in in one go.

**Comment by jheaff1** [comment]
> I could have sworn I put a reply on this PR but obviously didn't hit comment!
> I think a number of these changes are useful in their own rights so I've started creating PRs to merge them as separate PRs; hope you don't mind; once the framework changes are broken out then I think the meson changes can go in in one go.

No worries, thank you very much for taking the time to look at this!

**Comment by jsharpe** [comment]
@jheaff1 any ideas why this has started failing in CI?

**Comment by jheaff1** [comment]
It seems to be a remote caching issue, the curl test fails to launch because the OpenSSL shared libraries cannot be found. The issue may have been introduced in the following commit, where the OpenSSL build was changed to produce shared libs rather than static libs:

https://github.com/bazelbuild/rules_foreign_cc/commit/dd1e2850bd6523aad4d10ed71d171b78e4ace05a

**Comment by jheaff1** [comment]
We could test if it is a remote cache issue by building this branch without the remote cache. If it builds successfully, the issue must be due to remote caching

**Comment by jheaff1** [comment]
I had to disable the curl and OpenSSL tests for potentially the same reason in the sandboxed build 

https://github.com/bazelbuild/rules_foreign_cc/blob/9acbb356916760192d4c16301a69267fe44e6dec/.bazelci/config.yaml#L95

Perhaps we should disable these tests for the standalone build also?

**Comment by sfc-gh-mfickert** [comment]
Would it make sense to add a `setup_args` argument to the meson rule similar to `build_args` and `install_args` for the other steps?

**Comment by jheaff1** [comment]
> Would it make sense to add a `setup_args` argument to the meson rule similar to `build_args` and `install_args` for the other steps?

Sure, sounds like a good idea.

@jsharpe This PR seems to have stalled. any thoughts on https://github.com/bazelbuild/rules_foreign_cc/pull/986#issuecomment-1383234077?

**Comment by edison-moreland** [comment]
> Would it make sense to add a `setup_args` argument to the meson rule similar to `build_args` and `install_args` for the other steps?

This is a must if you’re cross compiling. I had to add `setup_args` to pass the cross file to meson

**Comment by jsharpe** [comment]
> I had to disable the curl and OpenSSL tests for potentially the same reason in the sandboxed build
> 
> https://github.com/bazelbuild/rules_foreign_cc/blob/9acbb356916760192d4c16301a69267fe44e6dec/.bazelci/config.yaml#L95
> 
> Perhaps we should disable these tests for the standalone build also?

Yes, if this is definitely down to the same issue, I'm happy with that for now; however the other reason I've yet to get round to reviewing this was due to the macOS CI breakage due to upgrades of the worker nodes. I've not checked if that's resolved yet?

**Comment by jheaff1** [comment]
> > I had to disable the curl and OpenSSL tests for potentially the same reason in the sandboxed build
> > https://github.com/bazelbuild/rules_foreign_cc/blob/9acbb356916760192d4c16301a69267fe44e6dec/.bazelci/config.yaml#L95
> > 
> > Perhaps we should disable these tests for the standalone build also?
> 
> Yes, if this is definitely down to the same issue, I'm happy with that for now; however the other reason I've yet to get round to reviewing this was due to the macOS CI breakage due to upgrades of the worker nodes. I've not checked if that's resolved yet?

I wasn’t aware of breakages due to worker node upgrades. Does the main branch build ok?

**Comment by jheaff1** [comment]
I've submitted #1016 to fix MacOS CI

File: toolchains/toolchains.bzl [hunk 1]
```diff
@@ -79,6 +80,17 @@ current_ninja_toolchain = rule(
     ],
 )
 
+current_meson_toolchain = rule(
+    implementation = _current_toolchain_impl,
+    attrs = {
+        "_toolchain": attr.string(default = str(Label("//toolchains:meson_toolchain"))),
```
Comment (jsharpe) [inline toolchains/toolchains.bzl:86]: I suspect this code will need to change to work with bzlmod but we can address that at a later date.

**Comment by jheaff1** [comment]
I’m not sure when I’ll have time to look at this, but I find the time I’ll sort it out 👍🏻

**Comment by jsharpe** [comment]
Can we just lift the zlib patch from https://github.com/conan-io/conan-center-index/blob/master/recipes/zlib/all/patches/1.2.13/0001-Fix-cmake.patch ?

**Comment by jheaff1** [comment]
> Can we just lift the zlib patch from https://github.com/conan-io/conan-center-index/blob/master/recipes/zlib/all/patches/1.2.13/0001-Fix-cmake.patch ?

Most likely

**Comment by inJeans** [comment]
Hello everyone. I am sure you are all very busy, but may I request you try and land this pull request? Please? It would be very helpful for me. Please :-)

**Comment by motiejus** [comment]
> Hello everyone. I am sure you are all very busy, but may I request you try and land this pull request? Please? It would be very helpful for me. Please :-)

This was discussed in `#cc` of bazelbuild.slack.com too if you wish more details: https://bazelbuild.slack.com/archives/CGA9QFQ8H/p1684329285418819?thread_ts=1683816175.149879&cid=CGA9QFQ8H

The original authors cleaned up the pull request and re-surfaced a new and smaller issue. They did more than they are supposed to. :) I guess this is up to someone else now to pick it up and finish.

I guess for anyone else seeing this thread — we should stop pinging the authors, just pick it up and finish it. I am personally at fault for briging it up myself (I originally had the bandwidth, but no longer), but let's be kind. :)

**Comment by inJeans** [comment]
I dont have access to that slack thread for some reason. But is the issue just the python builds?

**Comment by inJeans** [comment]
`@rules_foreign_cc_examples_third_party//python:python3_test              FAILED in 3 out of 3 in 0.2s`

**Comment by jheaff1** [comment]
> Hello everyone. I am sure you are all very busy, but may I request you try and land this pull request? Please? It would be very helpful for me. Please :-)

Because you asked so nicely... :)

@jsharpe CI is now building, could you please re-review?

File: examples/third_party/apr_util/BUILD.expat.bazel [hunk 2]
```diff
@@ -39,6 +39,7 @@ configure_make(
         "--without-docbook",
         "--without-examples",
         "--without-tests",
+        "CFLAGS='-fPIC'",
```
Comment (jsharpe) [inline examples/third_party/apr_util/BUILD.expat.bazel:42]: I don't love that this is required. Does `features=["pic"]` work for this?

File: examples/third_party/glib/BUILD.libffi.bazel [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+meson(
+    name = "libffi",
+    lib_source = ":all_srcs",
+    out_interface_libs = select({
+        "@platforms//os:windows": ["ffi.lib"],
+        "//conditions:default": [],
+    }),
+    out_lib_dir = select({
+        "@platforms//os:linux": "lib/x86_64-linux-gnu",
```
Comment (jsharpe) [inline examples/third_party/glib/BUILD.libffi.bazel:18]: This technically isn't constrained enough - it'd break on aarch64 linux for instance but it covers what is run in CI so I'm happy with this for now.

File: examples/third_party/mesa/BUILD.libdrm.bazel [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+filegroup(
+    name = "all_srcs",
+    srcs = glob(["**"]),
+)
+
+meson(
+    name = "libdrm",
+    lib_source = ":all_srcs",
+    out_lib_dir = "lib/x86_64-linux-gnu",
```
Comment (jsharpe) [inline examples/third_party/mesa/BUILD.libdrm.bazel:13]: Again this should probably be behind a select on the platform and arch but this is good enough for now.

File: foreign_cc/repositories.bzl [hunk 5]
```diff
@@ -12,6 +12,7 @@ def rules_foreign_cc_dependencies(
         cmake_version = "3.23.2",
         make_version = "4.4",
         ninja_version = "1.11.1",
+        meson_version = "0.63.0",
```
Comment (jsharpe) [inline foreign_cc/repositories.bzl:15]: There are now newer meson releases but lets land this first and then we can work on bumping the release to latest.

File: foreign_cc/repositories.bzl [hunk 6]
```diff
@@ -40,6 +41,8 @@ def rules_foreign_cc_dependencies(
         ninja_version: The target version of the ninja toolchain if `register_default_tools`
             or `register_built_tools` is set to `True`.
 
+        meson_version: The target version of the pkg_config toolchain if `register_built_tools` is set to `True`.
```
Comment (jsharpe) [inline foreign_cc/repositories.bzl:44]:
```suggestion
        meson_version: The target version of the meson toolchain.
```

**Review by jsharpe — APPROVED** [review]
@jheaff1 this is a mammoth effort! Thanks for all the work you've put into this, including all the previous bits that have built up to this in previous PRs that we've already landed!

Just one doc string that needs updating about meson_version but otherwise this LGTM.
A few observations but nothing that needs to block landing this; they can be addressed after this lands.

File: examples/third_party/apr_util/BUILD.expat.bazel (hunk 2 above)
Comment (jheaff1) [reply examples/third_party/apr_util/BUILD.expat.bazel:42]: I tried `features=["pic"]` and `-fPIC` was not added to the compiler command line args :(

File: foreign_cc/repositories.bzl (hunk 6 above)
Comment (jheaff1) [reply foreign_cc/repositories.bzl]: I've now resolved this

File: foreign_cc/repositories.bzl [hunk 7]
```diff
@@ -92,7 +96,7 @@ def rules_foreign_cc_dependencies(
     maybe(
         http_archive,
         name = "rules_python",
-        sha256 = "5fa3c738d33acca3b97622a13a741129f67ef43f5fdfcec63b29374cc0574c29",
-        strip_prefix = "rules_python-0.9.0",
-        url = "https://github.com/bazelbuild/rules_python/archive/refs/tags/0.9.0.tar.gz",
+        sha256 = "a3a6e99f497be089f81ec082882e40246bfd435f52f4e82f37e89449b04573f6",
+        strip_prefix = "rules_python-0.10.2",
+        url = "https://github.com/bazelbuild/rules_python/archive/refs/tags/0.10.2.tar.gz",
```
[the remark this answers is not in the document]
Comment (jheaff1) [reply foreign_cc/repositories.bzl]: I just went with the version of rules_python at the time i started doing all this work (about a year ago!). I've updated it to 0.22.0

[the remark this answers is not in the document]
Comment (jheaff1) [reply foreign_cc/repositories.bzl]: ah the failed CI build tells me that newer rules_python does not support bazel 4.0.0, only at least 5.4.0. I've now set the rules_python version used by rules_foreign_cc to 0.18.0, which was the last version to support bazel 4.0.0

[the remark this answers is not in the document]
Comment (jheaff1) [reply foreign_cc/repositories.bzl]: turns out 0.18.0 is also failing. I'll revert back to 0.10.2 if thats ok?

**Comment by jheaff1** [comment]
> @jheaff1 this is a mammoth effort! Thanks for all the work you've put into this, including all the previous bits that have built up to this in previous PRs that we've already landed!
> 
> Just one doc string that needs updating about meson_version but otherwise this LGTM. A few observations but nothing that needs to block landing this; they can be addressed after this lands.

Thank you very much. It only took me a year... 🤣

File: examples/third_party/mesa/BUILD.libdrm.bazel (hunk 4 above)
Comment (jheaff1) [reply examples/third_party/mesa/BUILD.libdrm.bazel:13]: I'll leave this as is

**Comment by inJeans** [comment]
Whoop! Thanks!! 🎉

**Comment by shixiong-jiang** [comment]
Hello, is meson supported today?
Could you plz add an example under the https://github.com/bazel-contrib/rules_foreign_cc/tree/main/examples? 

I plan to use it to compile https://github.com/systemd/systemd
Thanks a lot.

**Comment by jjmaestro** [comment]
> Hello, is meson supported today? Could you plz add an example under the https://github.com/bazel-contrib/rules_foreign_cc/tree/main/examples?

This PR already added an example: [`glib`](https://github.com/bazel-contrib/rules_foreign_cc/tree/main/examples/third_party/glib). Does that help?

**Comment by jsharpe** [comment]
Mesa is also built using meson: https://github.com/bazel-contrib/rules_foreign_cc/tree/main/examples/third_party/mesa

**Pull request merged**